### ISSUE TITLE
[ASCollectionNode] -referenceSizeForFooterInSectionASCollectionView crashes after updating to ASDK v.1.9.81

### ISSUE THREAD

**Issue by iwheelbuy** [facebookarchive/AsyncDisplayKit#1895]
**Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Supplementary node should exist.  Kind = UICollectionElementKindSectionFooter, indexPath = <NSIndexPath: 0x16f98ca0> {length = 2, path = 2 - 0}, collectionDataSource = <ASCollectionView: 0x17289000; baseClass = UICollectionView; frame = (0 0; 320 416); clipsToBounds = YES; gestureRecognizers = <NSArray: 0x16ec6e30>; layer = <<ASCollectionNode: 0x16ec9810>, layer = <_ASDisplayLayer: 0x16ebcdd0>>; contentOffset: {0, 45}; contentSize: {320, 853}> collection view layout: <UICollectionViewFlowLayout: 0x16eb4ba0>'**

The ViewController with ASCollectionView has successfully loaded. Then I moved to the next controller in UINavigationController stack. Everything was fine. Then I moved to the third controller in stack an get this weird crash. The third controller is not related to the first controller with ASCollectionView.

There is only one ASCollectionView in the project and there is only one place where supplementary node is used:

```
func collectionView(collectionView: ASCollectionView, nodeForSupplementaryElementOfKind kind: String, atIndexPath indexPath: NSIndexPath) -> ASCellNode {

        switch kind {
        case UICollectionElementKindSectionHeader where indexPath.section == 0:
            return self.header
        default:
            return ASCellNode()
        }
    }
```

ASDK v.1.9.74 and v.1.9.80 both work well without crash

**Comment by appleguy** [comment]
@erichoracek @Adlai-Holler it seems like this could be related to https://github.com/facebook/AsyncDisplayKit/pull/1773 .

I think the code is more correct than it was, but we should consider what we can learn from this and if there are additional incremental fixes that are possible without a full rearchitecture.  This was discussed in the ASDK team meeting today, and we'd love to prioritize it for the next couple weeks, but there is a challenging tradeoff with other 2.0 priorities / overall schedule.  It may need to be 2.1 before we can do the full rearchitecture.

**Comment by iwheelbuy** [comment]
The problem was related to `referenceSizeForFooterInSection`

```
/**
 * Asks the delegate for the size of the footer in the specified section.
 */
@available(iOS 6.0, *)
optional public func collectionView(collectionView: ASCollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForFooterInSection section: Int) -> CGSize
```

This method was implemeted with size (0.0;5.0)

and was implemented (as I already said)

```
func collectionView(collectionView: ASCollectionView, nodeForSupplementaryElementOfKind kind: String, atIndexPath indexPath: NSIndexPath) -> ASCellNode {

        switch kind {
        case UICollectionElementKindSectionHeader where indexPath.section == 0:
            return self.header
        default:
            return ASCellNode()
        }
    }
```

We removed `referenceSizeForFooterInSection` method and the crash disappeared.
Hope this info will help.

**Comment by Adlai-Holler** [comment]
@iwheelbuy We've made some major changes to supplementary views lately, especially for flow layouts.

In the current iteration, implement `- (ASSizeRange)collectionNode:(ASCollectionNode *)collectionNode sizeRangeForFooterInSection:(NSInteger)section` and return something nonzero (e.g. `ASSizeRangeUnconstrained`) and provide a supplementary node and it should work great. 

Please post here with an update if it's still not working.
